source/deimos/openssl/ossl_typ.d

```diff
@@ -1,263 +1,25 @@
-struct x509_crl_method_st;
-alias x509_crl_method_st X509_CRL_METHOD;
-struct x509_revoked_st;
-alias x509_revoked_st X509_REVOKED;
-struct X509_name_st;
-alias X509_name_st X509_NAME;
-struct X509_pubkey_st;
-alias X509_pubkey_st X509_PUBKEY;
-struct x509_store_st;
-alias x509_store_st X509_STORE;
-struct x509_store_ctx_st;
-alias x509_store_ctx_st X509_STORE_CTX;
-struct x509_lookup_st;
-alias X509_LOOKUP = x509_lookup_st;
-
-struct x509_object_st;
-alias X509_OBJECT = x509_object_st;
-struct x509_lookup_method_st;
-alias X509_LOOKUP_METHOD = x509_lookup_method_st;
-struct X509_VERIFY_PARAM_st;
-alias X509_VERIFY_PARAM = X509_VERIFY_PARAM_st;
-
-struct pkcs8_priv_key_info_st;
-alias pkcs8_priv_key_info_st PKCS8_PRIV_KEY_INFO;
-
-import deimos.openssl.x509v3;
-alias v3_ext_ctx X509V3_CTX;
-import deimos.openssl.conf;
-alias conf_st CONF;
-
-struct store_st;
-alias store_st STORE;
-struct store_method_st;
-alias store_method_st STORE_METHOD;
-
-struct ui_st;
-alias ui_st UI;
-struct ui_method_st;
-alias ui_method_st UI_METHOD;
-
-struct st_ERR_FNS;
-alias st_ERR_FNS ERR_FNS;
-
-struct engine_st;
-alias engine_st ENGINE;
-
-struct X509_POLICY_NODE_st;
-alias X509_POLICY_NODE_st X509_POLICY_NODE;
-struct X509_POLICY_LEVEL_st;
-alias X509_POLICY_LEVEL_st X509_POLICY_LEVEL;
-struct X509_POLICY_TREE_st;
-alias X509_POLICY_TREE_st X509_POLICY_TREE;
-struct X509_POLICY_CACHE_st;
-alias X509_POLICY_CACHE_st X509_POLICY_CACHE;
-
-alias AUTHORITY_KEYID_st AUTHORITY_KEYID;
-alias DIST_POINT_st DIST_POINT;
-alias ISSUING_DIST_POINT_st ISSUING_DIST_POINT;
-alias NAME_CONSTRAINTS_st NAME_CONSTRAINTS;
-
-  /* If placed in pkcs12.h, we end up with a circular depency with pkcs7.h */
-mixin template DECLARE_PKCS12_STACK_OF (type) { /* Nothing */ }
-//#define IMPLEMENT_PKCS12_STACK_OF!(type) /* Nothing */
-
-import deimos.openssl.crypto;
-alias crypto_ex_data_st CRYPTO_EX_DATA;
-/* Callback types for crypto.h */
-alias typeof(*(ExternC!(int function(void* parent, void* ptr, CRYPTO_EX_DATA* ad,
-					int idx, c_long argl, void* argp))).init) CRYPTO_EX_new;
-alias typeof(*(ExternC!(void function(void* parent, void* ptr, CRYPTO_EX_DATA* ad,
-					int idx, c_long argl, void* argp))).init) CRYPTO_EX_free;
-alias typeof(*(ExternC!(int function(CRYPTO_EX_DATA* to, CRYPTO_EX_DATA* from, void* from_d,
-					int idx, c_long argl, void* argp))).init) CRYPTO_EX_dup;
-
-import deimos.openssl.ocsp;
-struct ocsp_req_ctx_st;
-alias ocsp_req_ctx_st OCSP_REQ_CTX;
-/*struct ocsp_response_st;*/
-alias ocsp_response_st OCSP_RESPONSE;
-/*struct ocsp_responder_id_st;*/
-alias ocsp_responder_id_st OCSP_RESPID;
+// See PR https://github.com/openssl/openssl/pull/9333
+//
+// Note that we do print a deprecation even if the user depends on OpenSSL < 3.0.0
+// This is because the version of the bindings and the library are separated,
+// and updating the import in user code will be backward compatible
+// with older libraries.
+deprecated("Import `deimos.openssl.types` instead")
```

Comment:
I think we shouldn't do this.

Ideally, Deimos bindings should correspond 1:1 to C headers. This means that, given a C program that uses the targeted library, if you convert C syntax to D syntax (`#include` -> `import`, C casts -> D `cast` etc.), the program should compile and work as expected. Sticking to this as much as we can is, I think, an important quality of the Deimos project, because users of Deimos bindings know they won't have to adjust ported code to whatever whims each bindings project's authors decided to deviate from the C API.

I see in the linked OpenSSL PR that the old header file was left behind, silently redirecting to the new one:
https://github.com/openssl/openssl/pull/9333/files#diff-c65ac864946d4f5a545f90d2ca4083d3becd8a887a650db6147c6168a9e4ce2cR11

I think it would make the most sense if we did the same, without a deprecation. If OpenSSL stops shipping that header file, and we stop "supporting" all versions which shipped this header file, then we could add a deprecation or remove it entirely.

Reply:
I was hesitating between the current approach, deprecated in 3.0, and what you just described. Amended the PR accordingly.